### PR TITLE
👌 CLI: Computer/Code export `output_file` optional

### PR DESCRIPTION
**Pull request by GeigerJ2**
As the title states: The `output_file` argument for `verdi computer export [setup,config]` and `verdi computer export` is made optional by generating a default filename based on the respective label, if not specified.

I think this should be what people would probably call their files most of the time anyway.

### CODE REVIEW

File: src/aiida/cmdline/commands/cmd_code.py [hunk 1]
```diff
@@ -260,9 +260,14 @@ def export(code, output_file, sort):
         if value is not None:
             code_data[key] = str(value)
 
+    if output_file is None:
+        output_file = f"{code_data['label']}.yml"
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:264]: The option defines `Path(exists=False)` and so will error if the file exists. You probably want to apply the same check here. Even better though (since you will be reimplementing the logic of click) we try to make the default actual part of the option definition. Now since it depends on the code label, it cannot be a static default, but there is the concept of "contextual defaults" in click. See here for an example: https://github.com/aiidateam/aiida-core/blob/6db2f4060d4ece4552f5fe757c0f7d938810f4d1/src/aiida/cmdline/commands/cmd_user.py#L57

File: src/aiida/cmdline/commands/cmd_code.py [hunk 2]
```diff
@@ -260,9 +260,14 @@ def export(code, output_file, sort):
...
+
     with open(output_file, 'w', encoding='utf-8') as yfhandle:
         yaml.dump(code_data, yfhandle, sort_keys=sort)
 
+    echo.echo_success(f"Code<{code.pk}> {code.label} exported to file '{output_file}'.")
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:269]: Please stick to the convention of using backticks for variables
```suggestion
    echo.echo_success(f'Code<{code.pk}> {code.label} exported to file `{output_file}`.')
```

File: tests/cmdline/commands/test_code.py [hunk 3]
```diff
@@ -259,7 +259,7 @@ def test_code_duplicate_ignore(run_cli_command, aiida_code_installed, non_intera
 
 
 @pytest.mark.usefixtures('aiida_profile_clean')
-@pytest.mark.parametrize('sort', (True, False))
+@pytest.mark.parametrize('sort', ['--sort', '--no-sort'])
```
Comment (sphuber) [inline tests/cmdline/commands/test_code.py:262]: This change is not correct I believe. Now the test always runs with `--sort` since on line 285 you test for truthness of a string which is always `True`.

File: tests/cmdline/commands/test_code.py [hunk 4]
```diff
@@ -271,6 +271,16 @@ def test_code_export(run_cli_command, aiida_code_installed, tmp_path, file_regre
...
 
+    # Run export without explicit output_file path
+    # Don't provide output_file argument, but still export in tmp_path to avoid pollution of main repo when running test
+    cwd = os.getcwd()
+    os.chdir(tmp_path)
+    options = [str(code.pk)]
+    run_cli_command(cmd_code.export, options)
+    assert filepath.is_file
+    filepath.unlink()
+    os.chdir(cwd)
```
Comment (sphuber) [inline tests/cmdline/commands/test_code.py:282]: I think this is best accomplished by using the `chdir_tmp_path` fixture instead

File: tests/cmdline/commands/test_computer.py [hunk 5]
```diff
@@ -541,10 +541,40 @@ def test_computer_export_setup(self, tmp_path, sort):
...
+        computer_label = 'test_computer'
+        self.comp_builder.label = computer_label
+        self.comp_builder.transport = 'core.ssh'
+        comp = self.comp_builder.new()
+        comp.store()
+
+        # Export setup without explicit filepath
+        exported_filename = tmp_path / f'{computer_label}-setup.yml'
+        cwd = os.getcwd()
+        os.chdir(tmp_path)
```
Comment (sphuber) [inline tests/cmdline/commands/test_computer.py:556]: idem, use `chdir_tmp_path` instead

File: tests/cmdline/commands/test_computer.py [hunk 6]
```diff
@@ -541,10 +541,40 @@ def test_computer_export_setup(self, tmp_path, sort):
...
+        result = self.cli_runner(computer_export_setup, [comp.label])
+        assert result.exit_code == 0, 'Command should have run successfull.'
```
Comment (sphuber) [inline tests/cmdline/commands/test_computer.py:558]: This check is superfluous as it is done automatically by the `self.cli_runner` call

File: tests/cmdline/commands/test_computer.py [hunk 7]
```diff
@@ -541,10 +541,40 @@ def test_computer_export_setup(self, tmp_path, sort):
...
+        assert computer_label in result.output, 'Filename should be in terminal output but was not found.'
+        assert exported_filename.exists(), f"'{exported_filename}' was not created during export."
+        os.chdir(cwd)
+
+        # Export setup without explicit filepath
```
Comment (sphuber) [inline tests/cmdline/commands/test_computer.py:563]:
```suggestion
        # Export configuration without explicit filepath
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2 fully agree with the feature, just some comments on the implementation

File: src/aiida/cmdline/commands/cmd_code.py (hunk 2 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:269]: Hm, I agree with being consistent here. For that reason, I used the same setup as for the `echo.echo_success` of `verdi computer export`. So I'll go ahead and modify it there, as well.

File: tests/cmdline/commands/test_code.py (hunk 4 above)
Comment (GeigerJ2) [reply tests/cmdline/commands/test_code.py:282]: Done!

File: src/aiida/cmdline/commands/cmd_code.py (hunk 1 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:264]: Not sure `contextual_default` really is the best way here? It doesn't seem to be something that's implemented in `click` itself, but in the constructor of the `InteractiveOption` class:
https://github.com/aiidateam/aiida-core/blob/6db2f4060d4ece4552f5fe757c0f7d938810f4d1/src/aiida/cmdline/params/options/interactive.py#L50-L59

And I also only ever saw it used for these interactive options. Instead, I propose we make `OUTPUT_FILE` a globally available `option` under `cmdline/params/options/main.py`, and still set the default similarly as it is done in the implementation of my PR? In addition, I'd actually set `Path(exists=True)`, as I don't think it's a problem to silently overwrite these config YAML files (I'd change the label to the full `code@computer.yml` though).

Furthermore, `--output-file` is also used in `verdi node graph generate`, so I'd replace that there with the global `option` to not have three different ways of achieving the same thing, as `OUTPUT_FILE` also exists as a required, positional `argument` in `cmdline/params/arguments/main.py`. I'm not too happy with having it defined in two places eventually, but I'd still also keep it there, as that is used for `verdi archive create`, for which an automatic naming scheme is not as straightforward as for a `Computer` and `Code` (but happy to learn about a better alternative).

Pinging @mbercx as I mentioned this to him today.

File: tests/cmdline/commands/test_code.py (hunk 3 above)
Comment (GeigerJ2) [reply tests/cmdline/commands/test_code.py:262]: Good catch! I wanted to be consistent with the `['--sort', '--no-sort']` parametrization for `test_computer.py`. Though, now that I looked at it again, I'm not sure that it does what it's supposed to do even there... For the sake of this PR, it's probably best to just revert this change.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 1 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py:264]: TLDR: I wrote a bunch of separate comments, and then realized that really the first question to ask is to make sure you understand the difference between option and argument and whether you are proposing to change the output file _arguments_ for these export commands into output file _options_? If so, I don't think I agree and they should stay arguments. 

-----


You are right, the `contextual_default` is something that we implement, but you could use [`callbacks`](https://click.palletsprojects.com/en/8.1.x/options/#callbacks-for-validation) for the same purpose, which is a `click` native feature.

However, there is a bigger problem. These are both possibilities for _options_ but the current parameters you are changing are _arguments_, i.e., positional parameters, and they do not have the concept of callbacks.

> I propose we make OUTPUT_FILE a globally available option under cmdline/params/options/main.py

Since it is not an option but an argument, if we globalize it, it should go in `aiida.cmdline.params.arguments.main` instead, and there already is the `OUTPUT_FILE`. Or are you suggesting to change it from an argument into an option?

> In addition, I'd actually set Path(exists=True), as I don't think it's a problem to silently overwrite these config YAML files (I'd change the label to the full code@computer.yml though).

Here I don't agree. It could very well be that a user has a _different_ yaml in their CWD that is for a different code/computer but just happens to share the same name as the default that would be generated automatically. I think silently overriding here is bad, _especially_ if specifying the output name is optional.

> Furthermore, --output-file is also used in verdi node graph generate, so I'd replace that there with the global option to not have three different ways of achieving the same thing as OUTPUT_FILE also exists as a required, positional argument in cmdline/params/arguments/main.py. I'm not too happy with having it defined in two places eventually,

Note that here it _has_ to be an option. It actually used to be an argument, but we were forced to change it in order to allow multiple root nodes to be specified through arguments.
This demonstrates that an option and argument are not interchangeable solutions. Sometimes you need on or the other, or one or the other would be preferred.

For commands where the only argument would be an output file, I think it _should_ be an argument and usually it is required. But if we provide a default it technically becomes optional but I think it can be justified to keep it as an argument instead of changing it into an option.

File: tests/cmdline/commands/test_code.py (hunk 3 above)
Comment (sphuber) [reply tests/cmdline/commands/test_code.py:262]: > Though, now that I looked at it again, I'm not sure that it does what it's supposed to do even there...

It does use it correctly in `test_computer_export_setup`, although it also parametrizes `test_computer_export_config` but then doesn't appear to use it, so that would be a bug.

As for changing the code export test. You would simply add it directly to the options, e.g. something like:
```python
@pytest.mark.usefixtures('aiida_profile_clean')
@pytest.mark.parametrize('sort_option', ['--sort', '--no-sort'])
def test_code_export(run_cli_command, aiida_code_installed, tmp_path, file_regression, sort_option):
    ...
    options = [str(code.pk), str(filepath), sort_option]
    # options.append('--sort' if sort else '--no-sort')

    run_cli_command(cmd_code.export, options)
```

File: src/aiida/cmdline/commands/cmd_code.py (hunk 1 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:264]: The way I understand is that `option`s are specified via a single / double hyphen specifiers, while `argument`s are positional.

I'm not proposing we change them everywhere, and I don't really have an opinion / preference on what to make them. The problem, as you said, is that for arguments the `contextual_default` and `callback` (though, this I didn't know) features don't work, and that was the reason I considered turning it into an `option` instead. If setting the default can be achieved differently (e.g. via the logic I proposed initially), I'm fine with keeping them as (now optional) `argument`s.

@mbercx mentioned he'd like to have the same behavior as for `verdi node graph generate`, though, as the commands are different, I think it's fine that their API is also slightly different.

> Note that here it has to be an option. It actually used to be an argument, but we were forced to change it in order to allow multiple root nodes to be specified through arguments.

This I didn't know, thanks for providing the background (though, scope-creeping here, why not allow for multiple codes to be exported? :p)

Also fine for me to keep failing if the file already exists, so that the user is on the safe side 👍🏽

Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:264]: Yeah, I agree it's not very useful without that (and in general) :D

**Comment by sphuber** [comment]
@GeigerJ2 I would like to release today please and this is the last open PR. Would it be possible to wrap this up a.s.a.p.? Or shall we punt it?

**Comment by GeigerJ2** [comment]
We're on a group hike today, so unfortunately cannot work on it right now. I could wrap it up tomorrow morning first thing. Otherwise, if it cannot wait, feel free to release without, don't think it's that crucial.

**Comment by GeigerJ2** [comment]
Alright, so I think this is ready for a second round of reviews. I should have resolved your original comments, @sphuber. Though, there's no rush as the new version is already released. For all three commands (`verdi code export`, `verdi computer export [setup,config]`), I also added the `overwrite` option, and the commands fail if the files already exist and `overwrite` is `False`. In addition, I defined the `sort` option as a general option in `cmdline/params/options/main.py` as it's now being used in multiple places.

After taking a closer look at the tests again, I felt like the parametrization for the sorting was actually a bit unnecessary, so I removed it in the last commit (`db6259611`). It didn't really matter for the logic in the test function bodies, and just duplicated the number of tests being run in the (already very extensive) test suite. Therefore, I removed the parametrization, and just added an individual test for the sorting to each function. Similarly, for the `overwrite` option, I'm testing that the command fails when `overwrite` is `False`, and that the change is reflected in the output config file after modifying a `Code`/`Computer` instance and writing with `overwrite=True` (thanks to @mbercx for the discussion on testing). Also pinging @agoscinski as the implementations of the `Computer` export commands were written by him.

File: src/aiida/cmdline/commands/cmd_computer.py [hunk 8]
```diff
@@ -806,6 +804,10 @@ def computer_export_config(computer, output_file, user, sort):
         )
     try:
         computer_configuration = computer.get_configuration(user)
+        if output_file is None:
+            output_file = pathlib.Path(f'{computer.label}-config.yml')
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_computer.py:808]: You are not actually testing the case of the default filename. Please add this test for all commands

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2 . Fine with the implementation, just some minor implementation details. I am not sure I like the approach of the tests though where you are using a single test for all possible combinations. This tends to be very fragile as later tests in the function implicitly rely on some preconditions that were created (or not) by previous asserts and calls to the command. I think it is better to have separate test functions for the various options that are well separated. You are then forced to set up the necessary pre-conditions (for example pre-creating the output file in the case you are testing the user specifying a file that already exists) and it becomes also very clear when reading.

Also, I think using parametrization and using `file_regression` is a good thing. It makes for lean and easy to read tests. Now you are manually implementing the parametrization and reading/comparing outputs that is much more fragile and difficult to follow.

**Comment by GeigerJ2** [comment]
Thanks a lot for the review, @sphuber! Already a top-level answer, will resolve the issues soon: 

> I am not sure I like the approach of the tests though where you are using a single test for all possible combinations. This tends to be very fragile as later tests in the function implicitly rely on some preconditions that were created (or not) by previous asserts and calls to the command. I think it is better to have separate test functions for the various options that are well separated. You are then forced to set up the necessary pre-conditions (for example pre-creating the output file in the case you are testing the user specifying a file that already exists) and it becomes also very clear when reading.

Yeah, I was also not sure about that. As of now, the functions are very long. I'll split them up. Also, that's a good point that you bring up with some parts of the tests being reliant on previous calls of the function being tested, I'll definitely keep that consideration in mind!
 
> Also, I think using parametrization and using `file_regression` is a good thing. It makes for lean and easy to read tests. Now you are manually implementing the parametrization and reading/comparing outputs that is much more fragile and difficult to follow.

Using `file_regression` is good, I agree. In general, I wanted to avoid all the code in the test function body being run with the parametrization, e.g. `--sort/--no-sort`, even if it did not influence the actual part being tested, but doubling the number of tests being run. This will be resolved by splitting up the functions accordingly, then I will keep parametrization and `file_regression` where applicable.

Further, I also considered parametrizing all input parameters, e.g. `--sort/--no-sort` and no overwrite / `--overwrite`, but I don't think that makes sense, as the things I'm testing for don't necessarily depend on the values of these two parameters, and I would need to write a bunch of custom logic inside the test function body to differentiate the cases. So yeah, splitting up the test functions, using parametrization where applicable, and generating the states needed for testing without relying on the function being tested seems like a good way to move forward. Still new to this whole testing business :)

File: src/aiida/cmdline/commands/cmd_code.py [hunk 9]
```diff
@@ -234,34 +234,36 @@ def show(code):
...
 @arguments.CODE()
-@arguments.OUTPUT_FILE(type=click.Path(exists=False))
-@click.option(
-    '--sort/--no-sort',
-    is_flag=True,
-    default=True,
-    help='Sort the keys of the output YAML.',
-    show_default=True,
-)
+@arguments.OUTPUT_FILE(type=click.Path(exists=False), required=False)
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:237]: Done!

File: src/aiida/cmdline/commands/cmd_code.py [hunk 10]
```diff
@@ -234,34 +234,36 @@ def show(code):
...
         # If the attribute is not set, for example ``with_mpi`` do not export it, because the YAML won't be valid for
         # use in ``verdi code create`` since ``None`` is not a valid value on the CLI.
         if value is not None:
             code_data[key] = str(value)
 
-    with open(output_file, 'w', encoding='utf-8') as yfhandle:
-        yaml.dump(code_data, yfhandle, sort_keys=sort)
+    if output_file is None:
+        output_file = f"{code_data['label']}@{code_data['computer']}.yml"
+    output_file = Path(output_file)
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:259]: Done!

File: src/aiida/cmdline/commands/cmd_code.py [hunk 11]
```diff
@@ -234,34 +234,36 @@ def show(code):
...
+
+    if output_file.is_file() and not overwrite:
+        echo.echo_critical(f'File `{output_file}` already exists and overwrite set to {overwrite}.')
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:262]: Yes, good point!

File: src/aiida/cmdline/commands/cmd_computer.py [hunk 12]
```diff
@@ -769,6 +764,14 @@ def computer_export_setup(computer, output_file, sort):
         'prepend_text': computer.get_prepend_text(),
         'append_text': computer.get_append_text(),
     }
+
+    if output_file is None:
+        output_file = f'{computer.label}-setup.yml'
+    output_file = pathlib.Path(output_file)
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_computer.py:770]: Done!

File: tests/cmdline/commands/test_computer.py [hunk 13]
```diff
@@ -534,17 +532,42 @@ def test_computer_export_setup(self, tmp_path, sort):
...
+        with open(exported_setup_filename, 'r', encoding='utf-8') as fhandle:
+            content = fhandle.read()
+        assert 'core.local' in content
+
+        # Check unsorted file
+        result = self.cli_runner(
+            computer_export_setup, [comp.label, exported_setup_filename, '--overwrite', '--no-sort']
+        )
+        with open(exported_setup_filename, 'r', encoding='utf-8') as fhandle:
+            content = fhandle.read()
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply tests/cmdline/commands/test_computer.py:555]: Done!

File: src/aiida/cmdline/commands/cmd_code.py [hunk 14]
```diff
@@ -234,34 +234,36 @@ def show(code):
...
         if value is not None:
             code_data[key] = str(value)
 
-    with open(output_file, 'w', encoding='utf-8') as yfhandle:
-        yaml.dump(code_data, yfhandle, sort_keys=sort)
+    if output_file is None:
+        output_file = f"{code_data['label']}@{code_data['computer']}.yml"
+    output_file = Path(output_file)
+
+    if output_file.is_file() and not overwrite:
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:261]: Yeah, definitely quite the edge case, but ideally we capture it. Will put the logic in place!

File: tests/cmdline/commands/test_computer.py [hunk 15]
```diff
@@ -534,17 +532,42 @@ def test_computer_export_setup(self, tmp_path, sort):
...
+        print('CONTENT', content)
+        assert not content.startswith('append_text')
+        # This might break if the backend order changes (unlikely). Probably enough to just do the assert above?
+        assert content.startswith('label: test_computer_export_setup')
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply tests/cmdline/commands/test_computer.py:559]: Yeah, wasn't too happy about this either. I'm splitting the test now and using `file_regression` for the function that is using the `--sort/--no-sort` parametrization.

File: src/aiida/cmdline/commands/cmd_computer.py [hunk 16]
```diff
@@ -806,6 +804,10 @@ def computer_export_config(computer, output_file, user, sort):
         )
     try:
         computer_configuration = computer.get_configuration(user)
+        if output_file is None:
+            output_file = pathlib.Path(f'{computer.label}-config.yml')
+        if pathlib.Path(output_file).is_file() and not overwrite:
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_computer.py:809]: Done!

File: src/aiida/cmdline/commands/cmd_computer.py (hunk 8 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_computer.py:808]: Ah, yes. Forgot to test this for the `Computer`.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py]: I'm checking now for `is_dir()` and erroring out, irrespective of `overwrite`. In the case of `is_file()` and `overwrite=False`, providing the error message that points the user to the `overwrite` option.

File: tests/cmdline/commands/test_code.py [hunk 17]
```diff
@@ -291,6 +295,36 @@ def test_code_export(run_cli_command, aiida_code_installed, tmp_path, file_regre
...
+    with open(filepath, 'r', encoding='utf-8') as fhandle:
+        content = fhandle.read()
+    assert not content.startswith('append_text')
+    # This might break if the backend order changes (unlikely). Probably enough to just do the assert above?
+    assert content.startswith('label: code')
+
+    # Check default name
+    # Call explicitly here, rather than only for setup/tear down, as the other code above doesn't require the
+    # `chdir_tmp_path` fixture
+    chdir_tmp_path
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply tests/cmdline/commands/test_code.py]: I tried this out and when passing `chdir_tmp_path` as an argument to the test function, one ends up in the `tmp_path` even without calling it in the function body at all. So indeed, I think this should only be used for the setup via `    @pytest.mark.usefixtures`.

**Comment by GeigerJ2** [comment]
Still requires some clean-up. Hope I get around this tomorrow.

**Comment by GeigerJ2** [comment]
So `tests-presto` GH action is still failing and, frankly, I'm not sure how to best resolve it. In particular, the exported `Computer` config file created with the `--no-sort` option is still _sorted_, or, rather, different from the expected file of the `file_regression`. When `--no-sort` is used, the output file should just mirror the order in which the fields are added in the `src`. I assume this order is different, and depends on the SQL backend, so the obtained `no_sort` file is different from what the `file_regression` expects (even though, `safe_interval` is a `Transport`, not a `Computer` property, so it should not depend on the SQL backend anyway? If I understand correctly, a `Computer` is by default still an entry in the PSQL table, so also wondering how that changes in the SQLite case?).

As discussed with @khsrali, I'm inclined to just remove the `file_regression` for the `--no-sort` case, and just check for the general file existence and content instead, and then test for the sorted file exactly, possibly via `file_regression` (even though there's now no parametrization anymore). This is because `--no-sort` shouldn't mean it's the _reverse order as compared to `--sort`_, but rather that it is undefined / arbitrary (like a Python dict). I remember you hinting at that in the past, @sphuber, so just wondering what you think. Do you see a better option? Changing the backend implementation order for the tests here to pass seems overkill (and unnecessary).

**Comment by sphuber** [comment]
It is not dependent on the backend. The order for output that is based on pydantic models will be determined by the order of the fields as they are declared in the model. However, that is currently only the case for the `Code`. The `Computer` does not yet have a model and the export hardcodes the list. (This would be fixed by [this PR](https://github.com/aiidateam/aiida-core/pull/6255) btw which would add pydantic model for all ORM classes)

The test that is failing is for the _configuration_ of the computer though, and those values are taken from the `Transport` classes, which _also_ don't have a pydantic model yet. There the configuration is returned as a Python dictionary and there, as you correctly state, there is no real order. So my suspicion is that this test can also randomly fail for the normal test suite and has nothing to do with what storage plugin is used for the test profile.

I would simply just disable the `file_regression` test for the config export and just check that the command did not fail for `--no-sort` and that "some" output was generated. Just don't pay attention to the order as we don't care really what the order is.

File: src/aiida/cmdline/commands/cmd_code.py [hunk 18]
```diff
@@ -234,34 +236,35 @@ def show(code):
...
-@arguments.OUTPUT_FILE(type=click.Path(exists=False))
-@click.option(
-    '--sort/--no-sort',
-    is_flag=True,
-    default=True,
-    help='Sort the keys of the output YAML.',
-    show_default=True,
-)
+@arguments.OUTPUT_FILE(type=click.Path(exists=False, path_type=pathlib.Path), required=False)
+@options.OVERWRITE()
```
Comment (khsrali) [inline src/aiida/cmdline/commands/cmd_code.py:240]: This might be a bit of hassle and maybe unnecessarily. This is what happens to user: 
1) Hit the error `FileExistsError` first time, 
2) discover there is such option as overwrite
3) run again with `--overwrite`

Instead one could easily just index files:
For example if file `mycomputer.yaml` exists just produce `mycomputer_1.yaml` if that exists just do `mycomputer_2.yaml`

File: src/aiida/cmdline/commands/cmd_code.py [hunk 19]
```diff
@@ -234,34 +236,35 @@ def show(code):
...
         if value is not None:
             code_data[key] = str(value)
 
-    with open(output_file, 'w', encoding='utf-8') as yfhandle:
-        yaml.dump(code_data, yfhandle, sort_keys=sort)
+    try:
+        output_file = generate_validate_output_file(
+            output_file=output_file, entity_label=code.label, overwrite=overwrite, appendix=f'@{code_data["computer"]}'
+        )
+    except (FileExistsError, IsADirectoryError) as e:
```
Comment (khsrali) [inline src/aiida/cmdline/commands/cmd_code.py:262]: Normally you can always create files with same name as folders in OS. 
I would say just ignore and still make the file. 
The less unnecessarily raise, less annoyed user, no?

File: src/aiida/cmdline/commands/cmd_computer.py [hunk 20]
```diff
@@ -769,6 +765,14 @@ def computer_export_setup(computer, output_file, sort):
...
         'append_text': computer.get_append_text(),
     }
+
+    try:
+        output_file = generate_validate_output_file(
+            output_file=output_file, entity_label=computer.label, overwrite=overwrite, appendix='-setup'
+        )
+    except (FileExistsError, IsADirectoryError) as e:
+        echo.echo_critical(message=e)
+
```
Comment (khsrali) [inline src/aiida/cmdline/commands/cmd_computer.py:775]: Maybe I'm wrong, but `generate_validate_output_file` seems like a redundant function :) ?
It raises two **built-in** errors `IsADirectoryError` & `FileExistsError` that are being handled all together right here.
I mean `write_text` itself raises `FileExistsError` , and in my opinion `IsADirectoryError` doesn't need to be raised at all. 

Apart from that, also has `output_file` both as input and output. Feels a bit weird :)

File: tests/cmdline/commands/test_computer.py [hunk 21]
```diff
@@ -515,69 +516,156 @@ def test_show(self):
...
+
+        result = self.cli_runner(computer_export_config, [comp.label, exported_config_filename, sort_option])
         assert 'Success' in result.output, 'Command should have run successfull.'
         assert (
             str(exported_config_filename) in result.output
         ), 'Filename should be in terminal output but was not found.'
         assert exported_config_filename.exists(), f"'{exported_config_filename}' was not created during export."
+        # file regression test
+        content = exported_config_filename.read_text()
+        file_regression.check(content, extension='.yml')
```
Comment (khsrali) [inline tests/cmdline/commands/test_computer.py:616]: This is more of a curiosity than a comment. 

I haven't check how file regression test is done in `aiida` but from my previous experience what I found weird about `file_regression` is it's desire in replacing correct files. 
Suppose the code is doing something funny, and produces wrong results, in this case `file_regression` fails and replaces the correct file with the wrong one.
Then if you run again, it always pass. Because `wrong == wrong`. To be clear running twice always pass.

I discussed this yesterday with Julian, maybe @sphuber have comments how this behavior is being handled in `aiida`

**Review by khsrali — COMMENTED** [review]
Thanks @GeigerJ2 I added some annoying comments :)

File: src/aiida/cmdline/commands/cmd_code.py (hunk 18 above)
Comment (khsrali) [reply src/aiida/cmdline/commands/cmd_code.py:240]: btw, this is what my browser does when I download files. They used to raise file exist in the past, and aske user for a new file name :)

Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py:240]: I am personally not a fan of this to be honest. If the user explicitly specified the output file, I wouldn't want the command to silently change it. Of course you could get around the "silently" a bit by printing a warning that the message was changed, but this will not help in scripting use cases.

Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:240]: Thanks for the comments, @khsrali! There's no such thing as an annoying review :)

I see where you're coming from, but I'm not sure that I like the idea of automatically indexing the files, as that leads to the output name of the file being different from the label of the AiiDA entity, and I'd like those to be consistent. Suppose I run `verdi computer list` and I'm getting the output `mycomputer`. When I export it, I expect the file to be named `mycomputer.yaml`. Instead, if the file is called `mycomputer_1.yaml`, I might get confused (ofc we can inform the user, but still). Then, the user might still not be aware of the `--overwrite` flag, and then has to either run the command again using that, or `mv` his file to overwrite the original one or get the directory polluted with a bunch of files.

In the end, with all these commands that write to the disk, we have to make some decisions, e.g. file names, how to handle overwriting, etc., and there's no right or wrong, and people have different preferences. I don't have a strong opinion, but slightly prefer the current behavior. Let's see what @sphuber thinks.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 19 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py]: Here I definitely don't agree. If I accidentally specify the filename that happens to be a directory and the command just deletes the entire directory and overwrites it with a file, I'd be pretty pissed. I think the "annoyance" of having to change the output filename or specify `--overwrite` is way less than all the accidental loss of data.

File: src/aiida/cmdline/commands/cmd_computer.py (hunk 20 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_computer.py]: The other functionality of the function is to define the default value of the output file in case one hasn't been specified by the user. This is also why the function returns the output file

File: src/aiida/cmdline/commands/cmd_code.py (hunk 19 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py]: Personally, I would find it super confusing to have both, a file, and a directory with exactly the same name, even if it is theoretically possible. In any case, it should be very much an edge case, I don't expect users to have directories that end with `.yml`. And if they do, they should be notified of that, I think :D

File: tests/cmdline/commands/test_computer.py (hunk 21 above)
Comment (sphuber) [reply tests/cmdline/commands/test_computer.py]: That is not how the `file_regression` fixture works. When the reference file does not exist yet, it is created. However, if it already exists and the file generated by the tests is different, it is _not_ replaced but the test simply fails. Only if you run the tests with the `--force-regen` flag, will reference files be updated if they are different

Comment (GeigerJ2) [reply tests/cmdline/commands/test_computer.py]: That's weird, I never saw it overwriting files. From my experience, when adding a `file_regression`, it fails in the first run, as it doesn't have the file it's supposed to compare against, and creates it, which is also written in the output. Ofc, one could otherwise also add the expected file manually, but I found it quite convenient. Then, once the file is part of the repo, it is used for comparison, but not touched anymore, so the test fails if the implementation changes.

**Comment by GeigerJ2** [comment]
> It is not dependent on the backend. The order for output that is based on pydantic models will be determined by the order of the fields as they are declared in the model. However, that is currently only the case for the `Code`. The `Computer` does not yet have a model and the export hardcodes the list. (This would be fixed by [this PR](https://github.com/aiidateam/aiida-core/pull/6255) btw which would add pydantic model for all ORM classes)
> 
> The test that is failing is for the _configuration_ of the computer though, and those values are taken from the `Transport` classes, which _also_ don't have a pydantic model yet. There the configuration is returned as a Python dictionary and there, as you correctly state, there is no real order. So my suspicion is that this test can also randomly fail for the normal test suite and has nothing to do with what storage plugin is used for the test profile.
> 
> I would simply just disable the `file_regression` test for the config export and just check that the command did not fail for `--no-sort` and that "some" output was generated. Just don't pay attention to the order as we don't care really what the order is.

Thanks for the explanation, @sphuber, that makes a lot of sense! I removed the `file_regression` for the config export and just generally check for the content. In the sorted case, I'm doing it explicitly with `startswith`, which I think is fine, as we're also explicitly configuring it before, so we know what to expect.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 19 above)
Comment (khsrali) [reply src/aiida/cmdline/commands/cmd_code.py]: @sphuber I don't think that's ever possible,
`write_text()` is never going to delete your directory and replace a with a file.

I understand @GeigerJ2 's point that having a folder with `.yml` is pretty rare anyways. 
The point I wanted to make was why to specify this rare scenario anyways. But ok.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 18 above)
Comment (khsrali) [reply src/aiida/cmdline/commands/cmd_code.py:240]: I see your points, my concern was mainly when the user is not specifying an `output_name` so he doesn't really care... But no strong opinion.

**Comment by GeigerJ2** [comment]
> @GeigerJ2 I was just about to approve and merge this until I saw the last commit? Why are you changing that? `sort` is not a built-in keyword, only a method on an iterable, i.e. `[1, 2, 3].sort()` and not `sort([1, 2, 3])`. So having a variable called `sort` is fine and is not redefining anything.

Wanted to be extra sure to not cause any confusion ^^ I can revert it.

**Review by sphuber — APPROVED** [review]

**Comment by sphuber** [comment]
Thanks a lot @GeigerJ2
